### DISCUSSION TITLE
Multiple consumers on same channel; message consumption order?

### DISCUSSION THREAD

**Discussion by adrhc** [rabbitmq/rabbitmq-java-client#732] [Q&A] [answered]
Let's suppose I add consumer c1 processing queue q1 and consumer c2 processing q2 on the same channel (c1 declared before c2); the ack is manual and the prefetch count for c1 is pc1=5 while for c2 is pc2=8. Both queues are initially empty but at some same moment 2 parallel producers start sending each 800 messages at same constant sending rate, let's say a fast one.

I'm trying to find, for a java client, what threads are used to make things work.

q1) What thread is downloading the messages? How does it know how  many messages to get? especially if many same queue consumers exist with different prefetch count (on different channels)

q2) How the channel is dispatching the available messages to consumers? e.g. 3 messages available for c1 and 5 for c2

**Comment by michaelklishin** [comment, marked as answer]
This belongs to the Java client repository then.

You should not depend on any particular dispatch order of consumers, on a single channel or across multiple ones. It's like depending on a specific thread interleaving order.

All inbound protocol threads are read off the socket using the general socket API or NIO. Each carries a channel ID (number),
so connections know what channel to dispatch to.

Channels dispatch all inbound protocol methods to a JDK execution service (you can provide your own) in a way that guarantees that they are *dispatched* in the same order as received. Obviously actual processing of different frames can take a different amount of time and resources, so we cannot assume that they also complete in that order.

Consumers are looked up using delivery's consumer tag value and their handler methods are invoked in the aforementioned execution service context.

[Prefetch](https://rabbitmq.com/confirms.html#acknowledgement-modes) is a channel-level feature first and foremost, consumer-level prefetch is supported but rarely used. Regardless, this value is only used by RabbitMQ nodes to decide whether a delivery
can be sent at the next "tick" (a queue run in RabbitMQ server codebase parlance).  Prefetch does not change anything
about how client libraries work.

**Reply by adrhc** [reply]
> This belongs to the Java client repository then.

It seems for me you already put it in the right "category", please indicate me if that's isn't so.

Thank you for the details though I would like the discussion to focus more on the less documented things.

The part that baffles me relates to the bold words (only):  

> Each Channel **has its own dispatch thread**. For the most common use case of one Consumer per Channel, this means Consumers do not hold up other Consumers. If you have multiple Consumers per Channel be aware that a long-running Consumer may **hold up dispatch** of callbacks to other Consumers on that Channel.

My understanding is that, for 1 Consumer instance (hence 1 queue), if 2 messages are available (aka "in flight", according to [Channel Prefetch Setting (QoS)](https://www.rabbitmq.com/confirms.html#channel-qos-prefetch) vocabulary) than they'll be _dispatched_ serially/orderly; this would be possible in 2 ways: 
a) the Channel is running on some allocated thread, waiting on a blocking queue and calling the Consumer on each received item
b) the "in flight" blocking queue is managed outside the Channel; each message is handled to the Consumer to run on a thread; after thread work completion the next message is handled to the Consumer. 

While also considering this:

> Consumer may **hold up dispatch** of callbacks to other Consumers

I guess b) is the answer and additionally, that "in flight" blocking queue has to contain the messages for both Consumers (e.g. 2 queues).

Please clarify the above, thank you!

**Reply by michaelklishin** [reply]
I'm not sure I'm following the "1 Consumer instance, hence 1 queue" part. You can have any number of consumers on a queue, on any given number of channels, or reuse `Consumer` instances and consume from multiple queues.

Channels are not backed by threads. They depend on an object that performs the dispatch, and all channel methods are
called from the I/O thread IIRC (so it's easy to see why dispatching cannot happen right in that same thread).

This client is open source software, you are welcome to read the actual code:

 * [`ConsumerWorkService`](https://github.com/rabbitmq/rabbitmq-java-client/blob/main/src/main/java/com/rabbitmq/client/impl/ConsumerWorkService.java)
 * [`WorkPool`](https://github.com/rabbitmq/rabbitmq-java-client/blob/main/src/main/java/com/rabbitmq/client/impl/WorkPool.java) (this is the part that is most relevant for ordering and the only "rate limiting" part there is in the client)
 * [`ChannelN`](https://github.com/rabbitmq/rabbitmq-java-client/blob/main/src/main/java/com/rabbitmq/client/impl/ChannelN.java#L348-L350) that dispatches `basic.deliver` frames after receiving them from the connection

**Reply by adrhc** [reply]
With the new information I now reduce my not-understood part to:

> a **long-running Consumer** may hold up dispatch of callbacks to other Consumers

It seems from documentation that the actual Consumer processing (long-running Consumer) is getting in the dispatching way; how can that be true?

PS: I mean "1 Consumer instance, hence 1 queue" when the consumer instance is only once used for Channel.basicConsume
PS: I'll study the source code but with your answer that might go faster

**Reply by michaelklishin** [reply]
There is a hint in the documented purpose of `WorkPool` (IIRC). You cannot guarantee ordering if at some point, at least at the execution service dispatch level, deliveries on a channel are not processed sequentially. Deliveries on other channels don't have to be blocked since `WorkPool` only guarantees ordering for one channel.

If you want to really learn how the code works, read the code.

**Reply by michaelklishin** [reply]
I see JavaDoc refers to a design document. It's under https://github.com/rabbitmq/rabbitmq-java-client/tree/main/doc/channels and refers to "channels" in the general concurrency sense, not in the AMQP 0-9-1 channel sense (even if there are broad similarities).

**Reply by adrhc** [reply]
After analysing the code I conclude that indeed **a long-running Consumer may hold up dispatch of callbacks to other Consumers**:
`WorkPoolRunnable.run() = { ... for (Runnable runnable : block) { runnable.run(); } ... }`
In the above code each Runnable maps to a Consumer pertaining to the same Channel; hence there might be some consumers running in the same thread in certain heavy load conditions.

On the other hand **Each Channel has its own dispatch thread** doesn't seem right; in certain heavy load conditions there might be many pending WorkPoolRunnable instances cached in the newFixedThreadPool's task queue (more correctly said in ThreadPoolExecutor's queue) hence many threads processing Consumer(s) pertaining to same Channel might exist.

**Reply by michaelklishin** [reply]
Isn't that obvious, that if you use a thread pool and there are more runnables than available threads, some things would have to be delayed? Nothing specific to this client, its design or RabbitMQ.

**Reply by adrhc** [reply]
Well, it is; the fact that RMQ documentation bothers to specify it (though obvious) makes me think there's something I miss; and in fact it is something special about RMQ one could miss:
`WorkPoolRunnable.run() = { ... for (Runnable runnable : block) { runnable.run(); } ... }`
in certain conditions this code will run in the same thread multiple `Consumer.handleDelivery` calls.

On the other hand _Each Channel has its own dispatch thread_ part is still confusing; why **own** when in fact it's a thread allocated by `ExecutionService` hence not owned by the Channel?

**Reply by michaelklishin** [reply]
@adrhc you keep mentioning this "Each Channel has its own dispatch thread" line without mentioning where it's coming from. I could not find that line in this repo or the website one using GItHub search and `ag`/`grep` locally.

The Channels design doc is from 2010. Docs can easily get out of date in over a decade of changes. You are welcome to submit a PR with corrections.

**Reply by acogoluegnes** [reply]
Messages for a given channel are dispatched to the application code sequentially (one after the other), but there's no guarantee this happens in the same thread.

**Reply by adrhc** [reply]
@michaelklishin check [Receiving Messages by Subscription ("Push API")](https://www.rabbitmq.com/api-guide.html#consuming) which is a section from [Java Client API Guide](https://www.rabbitmq.com/api-guide.html); just search for "Each Channel has its own dispatch thread". Sorry, I usually read very carefully the documentation and I believe what is written in it hence _Channel has its **own** dispatch thread_ is very intriguing for me.

**Reply by michaelklishin** [reply]
OK, updated in https://github.com/rabbitmq/rabbitmq-website/commit/05fc4cc512beeefbb7c202090629a2569b9c6d08

**Comment by michaelklishin** [comment]
In your example, the order of message consumption by two consumers with different prefetch values should be considered undefined.
It won't be easy to reason about. Both will get deliveries in natural batches (assuming they acknowledge at the same rate) and
one will have more messages in flight than the other at most points in time.

You have never stated the end goal but if it is to build a priority mechanism for consumers on top of prefetch, this is not  going to
work well. Not only [consumer priority](https://rabbitmq.com/consumers.html#priority) and [single active consumer](https://rabbitmq.com/consumers.html#single-active-consumer) are more obvious choices for most cases, I personally find
any design that depends on priority of consumers on a single queue to be very questionable. Priorities of consumers on different
queues aren't much better but at least I can recall some cases where it made some sense.
